command/keyring/keyring.go

```diff
@@ -153,27 +172,41 @@ func (c *cmd) Run(args []string) int {
 	return 0
 }
 
-func (c *cmd) handleList(responses []*consulapi.KeyringResponse) {
-	for _, response := range responses {
-		pool := response.Datacenter + " (LAN)"
-		if response.Segment != "" {
-			pool += fmt.Sprintf(" [%s]", response.Segment)
-		}
-		if response.WAN {
-			pool = "WAN"
-		}
+func formatResponse(response *consulapi.KeyringResponse, keys map[string]int) string {
+	result := []string{
+		"",
+		poolName(response.Datacenter, response.WAN, response.Segment) + ":",
+		formatMessages(response.Messages),
+		formatKeys(keys, response.NumNodes),
+	}
+	return strings.Replace(strings.Join(result, "\n"), "\n\n", "\n", -1)
+}
 
-		c.UI.Output("")
-		c.UI.Output(pool + ":")
+func poolName(dc string, wan bool, segment string) string {
+	pool := fmt.Sprintf("%s (LAN)", dc)
+	if wan {
+		pool = "WAN"
+	}
+	if segment != "" {
+		segment = fmt.Sprintf(" [%s]", segment)
+	}
+	return fmt.Sprintf("%s%s", pool, segment)
+}
 
-		for from, msg := range response.Messages {
-			c.UI.Output(fmt.Sprintf("  ===> %s: %s", from, msg))
-		}
+func formatMessages(messages map[string]string) string {
+	result := []string{}
+	for from, msg := range messages {
+		result = append(result, fmt.Sprintf("  ===> %s: %s", from, msg))
+	}
+	return strings.Join(result, "\n")
```

Comment:
minor: this could be pre-allocated since we know the desired size: 

```go
result := make([]string, 0, len(messages))
```

Or, possibly even better, it could use a `strings.Builder` to avoid the temporary allocations required by `strings.Join`:

```go
b := new(strings.Builder)
for from, msg := range messages {
    b.WriteString(fmt.Sprintf("  ===> %s: %s\n", from, msg))
}
return b.String()
```

However this would add a trailing newline. So maybe it needs a strings.TrimSpace(), or maybe it is ok with the extra newline?


same for `formatKeys` I believe.

Reply:
Good call! The code is now only using the builder and that even helped to get rid of the replacing of `\n\n`.